dask/array/routines.py

```diff
@@ -531,10 +531,18 @@ def round(a, decimals=0):
 
 @wraps(np.unique)
 def unique(x):
-    name = 'unique-' + x.name
-    dsk = dict(((name, i), (np.unique, key)) for i, key in enumerate(x._keys()))
-    parts = Array._get(sharedict.merge((name, dsk), x.dask), list(dsk.keys()))
-    return np.unique(np.concatenate(parts))
+    x = x.ravel()
+
+    out = atop(np.unique, "i", x, "i", dtype=x.dtype)
+    out._chunks = tuple((np.nan,) * len(c) for c in out.chunks)
+
+    name = 'unique-' + out.name
+    dsk = {(name, 0): (np.unique, (np.concatenate, out._keys()))}
```

Comment:
I would have expected `out._keys()` to produce a numpy array of keys, which I would not expect the scheduler to understand well.

Reply:
Yeah I tried calling `tolist` on it initially, but then it complained that it couldn't call `tolist` on a `list` object.

Reply:
Do you have any ideas why this is? Should this be expected in this case?